### PR TITLE
HARP-8971: Fix wrong computation of fullFrameTime (hence fps)

### PR DESCRIPTION
**Pull request by ninok**
Previously the fullFrameTime was computed by subtracting the last frame start time from the current frame start time.
This was done to make sure we also measure the time we spend outside the current animation frame (i.e. geometry creation time).
Unfortunately with this approach we also measure everything the application/test is doing.

With this change the fullFrameTime is explicitly computed by adding the geometry creation time to the current frame.

Thank you for contributing to harp.gl!

Before requesting a pull request, please remember to check the following documents:
* [contribution guidelines](https://github.com/heremaps/harp.gl/blob/master/CONTRIBUTING.md)
* [coding style](https://github.com/heremaps/harp.gl/blob/master/CODINGSTYLE.md)

If you are adding new functionality we would highly appreciate if you can describe what is the capability you are adding and even better if you can add some examples. Please also remember to add tests for it.

# CI Check

Our bots will check whether your PR can be directly integrated into the mainline. We have some internal integration tests running on the background, our bots will inform you of the next steps and someone from our team will take a look and help if needed!

And please do not forget to sign-off your commit! You can read more about DCO [here](https://julien.ponge.org/blog/developer-certificate-of-origin-versus-contributor-license-agreements/). But, in short, you just need to use `git commit -s` or append `--signoff` when you are committing to the repo.

Happy contributing!

### CODE REVIEW

File: @here/harp-mapview/lib/Statistics.ts [hunk 1]
```diff
@@ -1017,55 +1017,51 @@ export class PerformanceStatistics {
...
+            this.currentFrame.setValue(
+                "gl.numPrograms",
+                webGlInfo.programs === null ? 0 : webGlInfo.programs.length
+            );
+        }
+    }
 
+    /**
+     * Add memory statistics to the current frame if available.
+     * @note Currently on supported on Chrome
```
Comment (kkostecki) [inline @here/harp-mapview/lib/Statistics.ts]: Typo: 

> only supported

File: @here/harp-mapview/lib/Statistics.ts [hunk 2]
```diff
@@ -1074,6 +1070,18 @@ export class PerformanceStatistics {
...
+    /**
+     * Stores the current frame events into the array of events and clears all values.
+     *
+     * @returns {boolean} Returns `false` if the maximum number of storable frames has been reached.
+     * @memberof PerformanceStatistics
+     */
+    storeAndClearFrameInfo(): boolean {
+        if (this.m_frameEvents.length >= this.maxNumFrames) {
+            return false;
+        }
```
Comment (kkostecki) [inline @here/harp-mapview/lib/Statistics.ts]: I guess it would be much better to store frames info in the round buffer, this way we never loose last frames information... Maybe this is a cause why we got some unpredictable results in some cases.
BTW: I know, it's not your code :).

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts [hunk 3]
```diff
@@ -286,6 +286,10 @@ export class SimpleTileGeometryLoader implements TileGeometryLoader {
             if (stats.enabled) {
                 const geometryCreationTime = PerformanceTimer.now() - now;
                 const currentFrame = stats.currentFrame;
+
+                // Account for the geometry creation in the current frame.
+                currentFrame.addValue("render.fullFrameTime", geometryCreationTime);
+
```
Comment (kkostecki) [inline @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: I wonder how much details we would like to spot... but reasonably we would also need to add the time for processing techniques via `TileGeometryCreator`. You may find it in `prepareForRender()` method in the same class.

File: @here/harp-mapview/lib/MapView.ts [hunk 4]
```diff
@@ -2709,30 +2709,32 @@ export class MapView extends THREE.EventDispatcher {
...
         const frameStartTime = time;
 
         RENDER_EVENT.time = time;
         this.dispatchEvent(RENDER_EVENT);
 
-        let currentFrameEvent: FrameStats | undefined;
-
         if (gatherStatistics) {
-            currentFrameEvent = stats.currentFrame;
+            currentFrameEvent = currentFrameEvent as FrameStats;
```
Comment (kkostecki) [inline @here/harp-mapview/lib/MapView.ts]: Although I understand this convention is look little bit awkward :).

Comment (ninok) [reply @here/harp-mapview/lib/MapView.ts]: Yeah, I was to lazy to add `!` everywhere

File: @here/harp-mapview/lib/Statistics.ts (hunk 2 above)
Comment (ninok) [reply @here/harp-mapview/lib/Statistics.ts]: true. for now I assume the buffer is big enough.

File: @here/harp-mapview/lib/geometry/TileGeometryLoader.ts (hunk 3 above)
Comment (ninok) [reply @here/harp-mapview/lib/geometry/TileGeometryLoader.ts]: Okay, I will check.

File: @here/harp-mapview/lib/Statistics.ts (hunk 1 above)
Comment (ninok) [reply @here/harp-mapview/lib/Statistics.ts]: done

File: @here/harp-mapview/lib/MapView.ts [hunk 5]
```diff
@@ -2917,15 +2922,25 @@ export class MapView extends THREE.EventDispatcher {
...
+
+            // Initialize the fullFrameTime with the frameRenderTime If we also create geometry in this frame
+            // this number will be increased in the TileGeometryLoader.
+            currentFrameEvent.setValue("render.fullFrameTime", frameRenderTime);
+
+            // Add THREE.js statistics
+            stats.addWebGLInfo(this.m_renderer.info);
 
-            PerformanceStatistics.instance.storeFrameInfo(this.m_renderer.info);
+            // Add memory statistics (FIXME: )
```
Comment (nzjony) [inline @here/harp-mapview/lib/MapView.ts]: What exactly needs to be fixed here? It is unclear to me.

Comment (ninok) [reply @here/harp-mapview/lib/MapView.ts]: ups, forgot to actually write it down.

File: @here/harp-mapview/lib/MapView.ts [hunk 6]
```diff
@@ -2713,30 +2713,35 @@ export class MapView extends THREE.EventDispatcher {
...
         }
+
         ++this.m_frameNumber;
+        const frameStartTime = time;
 
+        let currentFrameEvent: FrameStats | undefined;
         const stats = PerformanceStatistics.instance;
         const gatherStatistics: boolean = stats.enabled;
+        if (gatherStatistics) {
+            currentFrameEvent = stats.currentFrame;
```
Comment (nzjony) [inline @here/harp-mapview/lib/MapView.ts]: why was this moved here, can't it be part of the if(gatherStat...) on line 2742?

File: @here/harp-mapview/lib/Statistics.ts (hunk 2 above)
Comment (nzjony) [reply @here/harp-mapview/lib/Statistics.ts]: this seems to be able to store 1000 frames, at 60 fps, that is roughly 15 seconds worth of data, for a flyover test, that may not be enough?

File: @here/harp-mapview/lib/MapView.ts [hunk 7]
```diff
@@ -2713,30 +2713,35 @@ export class MapView extends THREE.EventDispatcher {
         if (this.m_drawing) {
             return;
         }
+
         ++this.m_frameNumber;
+        const frameStartTime = time;
```
Comment (nzjony) [inline @here/harp-mapview/lib/MapView.ts]: why do we have an alias for the time? Why not just rename the parameter ?

Comment (ninok) [reply @here/harp-mapview/lib/MapView.ts]: good point. I just copy pasted stuff around. Will just rename the param.

File: @here/harp-mapview/lib/Statistics.ts (hunk 2 above)
Comment (ninok) [reply @here/harp-mapview/lib/Statistics.ts]: I think the flyover test is fetching the statistics after every frame, so the buffer will not overflow but I will double check

File: @here/harp-mapview/lib/MapView.ts (hunk 6 above)
Comment (ninok) [reply @here/harp-mapview/lib/MapView.ts]: I think the perf test gathers the statistics in the `RENDER_EVENT. Will also double check this.

File: @here/harp-mapview/lib/Statistics.ts (hunk 2 above)
Comment (ninok) [reply @here/harp-mapview/lib/Statistics.ts]: @nzjony : Actually your suspicion was correct. By default we store 1000 frames but flyover renders 5000 without fetching statistics in between. What is also interesting is, that m_frameEvents is actually a ring buffer but we just stop putting statistics into it.

File: @here/harp-mapview/lib/MapView.ts (hunk 6 above)
Comment (ninok) [reply @here/harp-mapview/lib/MapView.ts]: the perftests are listening to the `AfterRender` event.So I move the dispatching of the render event to the very top (and also don't count it into the frame time b/c it's not rendering but application code)

File: @here/harp-mapview/lib/Statistics.ts (hunk 2 above)
Comment (ninok) [reply @here/harp-mapview/lib/Statistics.ts]: After looking again into our code and due to hints from @CraigBerlin I found out that the `PerformanceStatistics` are not really following the singleton pattern, i.e. `PerformanceStatistics.instance` will be recreated whenever a new `PerformanceStatistics` object is instantiated.
